### PR TITLE
FEATURE: implements topic_required_words script

### PR DESCRIPTION
**Pull request by jjaffeux**
This commit also introduces the topic trigger which let's an automation mark a topic has "automation-able" (through a custom field), scripts and other utilities can then check if a topic is automation-able before applying any logic on it.

### CODE REVIEW

**Comment by jjaffeux** [comment]
Capture d’écran 2021-03-30 à 11 10 33

File: app/lib/discourse_automation/triggers/topic.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        previous_topic = Topic.find_by(id: previous_topic_id)
+
+        if previous_topic
+          previous_topic.custom_fields.delete('discourse_automation_id')
+          previous_topic.save!
+        end
+      end
+
+      if topic_id
+        topic = Topic.find(topic_id)
```
Comment (ZogStriP) [inline app/lib/discourse_automation/triggers/topic.rb:20]: I believe the `find` method will trigger an exception if the topic doesn't exist. I think you want `find_by(id: ` instead.

File: plugin.rb [hunk 2]
```diff
@@ -121,6 +123,43 @@ class Engine < ::Rails::Engine
...
+  reloadable_patch do
+    require 'post'
+
+    class ::Post
+      validate :discourse_automation_topic_required_words
+
+      def discourse_automation_topic_required_words
+        if topic.custom_fields['discourse_automation_id'].present?
+          automation = DiscourseAutomation::Automation.find(topic.custom_fields['discourse_automation_id'])
+          if automation && automation.script == 'topic_required_words'
```
Comment (ZogStriP) [inline plugin.rb:138]:
```suggestion
          if automation&.script == 'topic_required_words'
```

File: plugin.rb [hunk 3]
```diff
@@ -121,6 +123,43 @@ class Engine < ::Rails::Engine
...
+            words = automation.fields.find { |field| field.name == 'words' }
+
+            return if !words
+
+            words = words.metadata['list']
+
+            if !words.blank?
```
Comment (ZogStriP) [inline plugin.rb:145]:
```suggestion
            if words.present?
```

File: plugin.rb [hunk 4]
```diff
@@ -121,6 +123,43 @@ class Engine < ::Rails::Engine
...
+              includes_at_least_one_word = false
+
+              words.each do |word|
+                if raw.include?(word)
+                  includes_at_least_one_word = true
+                end
+              end
```
Comment (ZogStriP) [inline plugin.rb:152]:
```suggestion
              includes_at_least_one_word = words.any? { |word| raw.include?(word) }
```

File: plugin.rb [hunk 5]
```diff
@@ -121,6 +123,43 @@ class Engine < ::Rails::Engine
...
+
+              if !includes_at_least_one_word
```
Comment (ZogStriP) [inline plugin.rb:154]:
```suggestion
              if words.none? { |word| raw.include?(word) }
```

**Comment by jjaffeux** [comment]
@ZogStriP yes much better thanks :+1:

**Pull request merged**